geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java

```diff
@@ -0,0 +1,216 @@
+
+public class ZAddIncrOptionDUnitTest {
+  @Rule
+  public ExecutorServiceRule executor = new ExecutorServiceRule();
+
+  @Rule
+  public RedisClusterStartupRule clusterStartUp = new RedisClusterStartupRule(4);
+
+  private JedisCluster jedis;
+  private List<MemberVM> servers;
+  private static final String sortedSetKey = "key";
+  private final String baseMemberName = "member";
+  private final int setSize = 1000;
+  private final double increment1 = 355.681000005;
+  private final double increment2 = 9554257.921450001;
+  private final double total = increment1 + increment2;
+
+  @Before
+  public void setup() {
+    MemberVM locator = clusterStartUp.startLocatorVM(0);
+    int locatorPort = locator.getPort();
+    MemberVM server1 = clusterStartUp.startRedisVM(1, locatorPort);
+    MemberVM server2 = clusterStartUp.startRedisVM(2, locatorPort);
+    MemberVM server3 = clusterStartUp.startRedisVM(3, locatorPort);
+    servers = new ArrayList<>();
+    servers.add(server1);
+    servers.add(server2);
+    servers.add(server3);
+
+    int redisServerPort = clusterStartUp.getRedisPort(1);
+
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, redisServerPort), REDIS_CLIENT_TIMEOUT);
+  }
+
+  @After
+  public void tearDown() {
+    jedis.close();
+  }
+
+  @Test
+  public void zAddWithIncrOptionCanAddAndIncrementScoresConcurrently() {
+    new ConcurrentLoopingThreads(setSize,
+        (i) -> doZAddIncr(i, increment1, total, true),
+        (i) -> doZAddIncr(i, increment2, total, true)).run();
+
+    assertThat(jedis.zcard(sortedSetKey)).isEqualTo(setSize);
+    verifyZScores();
+  }
+
+  private void verifyZScores() {
+    for (int i = 0; i < setSize; i++) {
+      assertThat(jedis.zscore(sortedSetKey, baseMemberName + i)).isEqualTo(total);
+    }
+  }
+
+  private void doZAddIncr(int i, double increment, double total, boolean isConcurrentExecution) {
+    Object result =
+        jedis.sendCommand(sortedSetKey, Protocol.Command.ZADD, sortedSetKey, "INCR",
+            Coder.doubleToString(increment), baseMemberName + i);
+    if (isConcurrentExecution) {
+      assertThat(Coder.bytesToDouble((byte[]) result)).isIn(increment, total);
+    } else {
+      assertThat(Coder.bytesToDouble((byte[]) result)).isEqualTo(total);
+    }
+  }
+
+  @Test
+  public void zAddWithIncrOptionCanIncrementScoresAfterPrimaryShutsDown() {
+    doZAddIncrForAllMembers(increment1, increment1);
+
+    stopNodeWithPrimaryBucketOfTheKey(false);
+
+    doZCardWithRetries();
+    doZAddIncrForAllMembers(increment2, total);
+    verifyZScores();
+  }
+
+  private void doZAddIncrForAllMembers(double increment1, double increment12) {
+    for (int i = 0; i < setSize; i++) {
+      doZAddIncr(i, increment1, increment12, false);
+    }
+  }
+
+  private void doZCardWithRetries() {
+    int maxRetryAttempts = 10;
+    int retryAttempts = 0;
+    while (!zCardWithRetries(retryAttempts, maxRetryAttempts)) {
+      retryAttempts++;
+    }
+  }
```

Comment:
It might be a bit more straightforward to take out the extra `zCardWithRetries` method and just do the following:
```
  private void doZCardWithRetries() {
    int maxRetryAttempts = 10;
    int retryAttempts = 0;

    while (retryAttempts < maxRetryAttempts) {
      long memberSize;
      try {
        memberSize = jedis.zcard(sortedSetKey);
      } catch (JedisClusterMaxAttemptsException e) {
          retryAttempts++;
          if(retryAttempts == maxRetryAttempts) {
            throw e;
          }
          continue;
      }
      assertThat(memberSize).isEqualTo(setSize);
      break;
    }
  }
 ```
 ...but that might just be a personal preference.